int32 success_flag
int32 average_error